### DISCUSSION TITLE
Improving the experience for PR contributors

### DISCUSSION THREAD

**Discussion by Jermolene** [TiddlyWiki/TiddlyWiki5#5484] [General]
Over in https://github.com/Jermolene/TiddlyWiki5/pull/4681#issuecomment-775052141, @rubaboo made an important comment that I thought was worth pulling out into its own discussion:

> @Jermolene,
> I've been tracking the project on and off since late beta (and using TW2 for a number of years prior). Over the years there have been some incredibly talented contributors that came and went. I would like to think that they moved on due to their own reasons, but I'm afraid some of them did so due to unsatisfying experience with their PRs. Just saying. My own experience has been overwhelmingly positive, due to the tiny size and scope of the few PRs that I've managed to submit in the past.

I'm horribly aware that the process can be frustrating for contributors. I try to engage properly with every PR, and give coherent feedback and explain my thinking, but am frequently behind with the volume of work and undoubtedly some PRs get neglected for too long.

I see a few different issues:

* For me, and perhaps counter intuitively, merging a PR is generally significantly more work than if I had made the modification myself. The kind of detailed reply (like this one) that I give on GitHub often takes 20-30 minutes, a fairly significant chunk of the time that I have available to work on the project. I'm not complaining, I find it helpful to get my thoughts in order, but it is surprisingly time consuming
* In the past I have sometimes merged PRs that I have subsequently deeply regretted. Often I did so because of a sense of obligation towards the contributor. However that usually leaves permanent scarring on the core code because of our backwards compatibility constraints. So I've learned to be moderately cautious, and not to merge PRs if there is doubt
* Sometimes people invest a large amount of effort on PRs without prior consultation. In those situations it becomes extremely difficult if I have reservations about merging it
* TiddlyWiki has aspirations to stand the test of time, and still be useful for as long as possible. So I take a long view, thinking in terms of decades. That means that there is a sense in which time should and does move relatively slowly for the core. That's why I frequently suggest that proposed changes be implemented as plugins if possible in the first instance
* The unmerged PRs are an obvious eye sore for visitors to the project, but GitHub doesn't draw so much attention to the large number of PRs that do get processed and merged in a reasonably timely manner (relative to our release cycle)
* Quite a few of the PRs we get are not of the highest quality, even sometimes from regular contributors. (My pet bugbear is PRs that are not self contained, and require reading through a Google Groups discussion to understand). I generally still try to engage with such PRs but it makes more work for me, hits my motivation, and slows things down. In an ideal world, PRs would be put together with primary consideration for ease of merging. That means making a detailed description, rationale, screenshots, and anything  else to help people understand the proposal
* It should go without saying, but if a contributor really wants their stuff merged then they need to make it as easy as possible by including clear descriptions, rationales and documentation
* I do intend to move TW5 into a GitHub organisation which will give us finer control over permissions so that others can merge PRs. But I don't think that that's really the problem: we're not constrained by how quickly I can click the merge button, but by how quickly me or anyone else can provide feedback. I'm incredibly grateful for the assistance that I get in that regard but I can't help noticing that, for example, I seem to be the only one who gives feedback about coding style violations

We're all fragile human beings, and we're working together in a way that hasn't been done before in history. No doubt for all of us it is sometimes a struggle to maintaining our motivation, and contain our frustration when things go wrong. I can't resolve all those issues for everyone or even just for me, so I try to treat people and their work with respect, and engage with everyone as much as I can. I try to share my thinking so that people have a chance to understand things from my perspective.

I'm interested in the views and experiences of others, and particularly in actionable suggestions for improvements to the process.

**Comment by rubaboo** [comment]
1. Just reading the above helps, to explain delays and to set expectations. I'd say put it right in `contributing.md` in some condensed form.
2. `contributing.md` needs to be expanded and prominently featured in the `readme.md`. Not everyone will think to read developers documentation before submitting small fixes and typos.
3. Make rules. Put a short version in `contributing.md` and make it a required reading before submitting anything. 

- One feature change per PR
- Self-contained
- Well documented and sufficiently justified
- Minimal amount of commits, meaningfully atomized
- Keep the coding style
- Read developers documentation
- Consult prior to making a large one

4.  Enforce the rules. Rather than spending time explaining things, just point out which rules the PR is violating, and let your helpers take over if the submitter needs more guidance.
5. I'm sure many will disagree, but I would not mind an occasional change that is not backward compatible. Better than permanent scarring.
6. There is no way around this one. For PRs that cannot be merged promptly for whatever reason, there needs to be something in place to ensure they do not fall off the plate. Something other than the submitter's continued enthusiasm.
7. Last but not least - please do keep merging to yourself for as long as you can. 🙂 Regardless of what I previously said, it is your constraint, vision, and commitment that allowed me to keep using the "bleeding edge" version, with only one or two hiccups over the years.

**Reply by clutterstack** [reply]
As an inexperienced contributor, I agree with much of the above from @rubaboo (and don't necessarily disagree with any of it). From my point of view, it seems fair (because necessary) to expect a bit more effort and resourcefulness of contributors in order for the resources available to review them to match the rate of incoming.

If a common first-pass response to a PR were more like a multiple-choice problem than an essay, would that put contributors off? If it came with friendly, actionable boilerplate? If this is an advertised default, I think it's a reasonable trade-off.

**Reply by Jermolene** [reply]
Thanks @rubaboo @clutterstack I've made some changes to `contributing.md` here: c976aad5e000332bb21d3c2a05f52e769f21ccb0

**Comment by linonetwo** [comment]
Just want to comment on this:

> The kind of detailed reply (like this one) that I give on GitHub often takes 20-30 minutes

You can try speech-to-text, it is very efficient. My recent tiddlers are mostly roughly done by using Sougou speech input (for Chinese input), and format them later. This can input tons of information from my brain to the computer in a small amount of time, and reduce the fatigue that typing long text gives me.
